src/main/scala/org/constellation/consensus/ResolutionService.scala

```diff
@@ -0,0 +1,109 @@
+package org.constellation.consensus
+
+import java.util.concurrent.TimeUnit
+
+import akka.actor.ActorRef
+import akka.util.Timeout
+import org.constellation.Data
+import org.constellation.primitives.{APIBroadcast, Schema}
+import org.constellation.primitives.Schema.{CheckpointBlock, CheckpointCacheData, SignedObservationEdgeCache}
+
+import scala.annotation.tailrec
+import scala.concurrent.{ExecutionContext, Future}
+
+object ResolutionService {
+  implicit val timeout: Timeout = Timeout(5, TimeUnit.SECONDS)
+
+  case class ResolutionStatus(
+                               cb: CheckpointCacheData,
+                               resolvedParents: Seq[(String, Option[SignedObservationEdgeCache])],
+                               unresolvedParents: Seq[(String, Option[SignedObservationEdgeCache])]
+                             )
+
+  /**
+    * TODO: Need to include signatories ABOVE this checkpoint block later in the case of signature decay. Add to SignedObservationEdgeCache
+    *
+    * @param dao
+    * @param cb
+    * @return
+    */
+  def downloadAncestry(dao: Data, cb: CheckpointBlock) = {
+    val observers: Set[Schema.Id] = cb.signatures.map {
+      _.toId
+    }.toSet
+    dao.queryMissingResolutionData(cb.baseHash, observers)
+  }
+
+  /**
+    *
+    * @param dao
+    * @param unresolvedChildren
+    */
+  def reprocessUnresolvedChildren(dao: Data, unresolvedChildren: Seq[CheckpointBlock])(implicit executionContext: ExecutionContext) = unresolvedChildren.foreach { unresolvedChild =>
+    EdgeProcessor.handleCheckpoint(unresolvedChild, dao, true) //Todo log exception add retry
+  }
+
+  /**
+    *
+    * @param dao
+    * @param cb
+    */
+  def reprocessUnresolvedParents(dao: Data, cb: CheckpointBlock): (String, Option[SignedObservationEdgeCache]) => Unit = {//Todo test me
```

Comment:
I think this would be clearer if you just use two parameter lists. (dao...)(missingParentHash...)

Reply:
Yeah, will just lift over option